In [1]:
%load_ext sql
%sql mysql+pymysql://root:mummypapa@localhost/sakila

In [60]:
%%sql
-- Q1. Create a table called employees
CREATE TABLE employees (
    emp_id INT NOT NULL PRIMARY KEY,
    emp_name VARCHAR(100) NOT NULL,
    age INT CHECK (age >= 18),
    email VARCHAR(255) UNIQUE,
    salary DECIMAL DEFAULT 30000
);

 * mysql+pymysql://root:***@localhost/sakila
0 rows affected.


[]

Q2 Purpose of Constraints in Database
Constraints in databases ensure the accuracy and reliability of data. By enforcing rules on the data, constraints help maintain data integrity and prevent invalid data from being added or modified. Here are common types of constraints:

NOT NULL: Ensures that a column cannot contain NULL values. Example: An emp_name column may be set to NOT NULL to ensure every employee has a name.
PRIMARY KEY: Uniquely identifies each row in a table, and it implicitly includes NOT NULL. Example: An emp_id column as a primary key ensures every employee has a unique identifier.
UNIQUE: Ensures all values in a column are unique. Example: An email column with a UNIQUE constraint prevents duplicate email addresses.
CHECK: Ensures a specified condition is met before data can be added or updated. Example: The age column has a CHECK constraint to allow only values of 18 or older.
FOREIGN KEY: Establishes a link between the data in two tables, enforcing referential integrity. Example: A department_id column in the employees table might refer to a department_id in a departments table.

Q3 
The NOT NULL constraint ensures that a column cannot have a NULL value. Applying NOT NULL is useful for required fields, such as emp_name in the employees table, to ensure each employee has a name. A primary key uniquely identifies each row, and by definition, it cannot contain NULL values because NULLs would violate the uniqueness rule. If a primary key contained NULL, it would lose its ability to uniquely identify rows.

In [62]:
%%sql
-- Q4 Add a constraints
ALTER TABLE employees
ADD CONSTRAINT age_check CHECK (age >= 18);


 * mysql+pymysql://root:***@localhost/sakila
0 rows affected.


[]

In [64]:
%%sql
-- Q4 Remove a Constraints
ALTER TABLE employees
DROP CONSTRAINT age_check;


 * mysql+pymysql://root:***@localhost/sakila
0 rows affected.


[]

Q5. Consequences of Violating Constraints

Attempting to insert, update, or delete data that violates constraints results in an error and prevents the operation from completing. This maintains the integrity of the data. For example, if we try to insert an employee with an age of 16, a CHECK constraint violation error will occur.

Q6. 
ALTER TABLE products
ADD CONSTRAINT pk_product_id PRIMARY KEY (product_id);
###### ALTER TABLE products MODIFY price DECIMAL(10, 2) DEFAULT 50.00;


Q7 SELECT students.student_name, classes.class_name
FROM students
INNER JOIN classes ON students.class_id = classes.class_id;


Q8 SELECT orders.order_id, customers.customer_name, products.product_name
FROM products
LEFT JOIN orders ON products.product_id = orders.product_id
LEFT JOIN customers ON orders.customer_id = customers.customer_id;


Q9 SELECT products.product_id, products.product_name, SUM(sales.amount) AS total_sales_amount
FROM products
INNER JOIN sales ON products.product_id = sales.product_id
GROUP BY products.product_id, products.product_name;


Q.10 SELECT 
    orders.order_id, 
    customers.customer_name, 
    SUM(order_items.quantity) AS quantity_ordered
FROM 
    orders
INNER JOIN 
    customers ON orders.customer_id = customers.customer_id
INNER JOIN 
    order_items ON orders.order_id = order_items.order_id
GROUP BY 
    orders.order_id, customers.customer_name;


# SQL Commands

In [3]:
%%sql
show tables

 * mysql+pymysql://root:***@localhost/sakila
24 rows affected.


[('actor',),
 ('actor_info',),
 ('address',),
 ('category',),
 ('city',),
 ('country',),
 ('customer',),
 ('customer_list',),
 ('employees',),
 ('film',),
 ('film_actor',),
 ('film_category',),
 ('film_list',),
 ('film_text',),
 ('inventory',),
 ('language',),
 ('nicer_but_slower_film_list',),
 ('payment',),
 ('rental',),
 ('sales_by_film_category',),
 ('sales_by_store',),
 ('staff',),
 ('staff_list',),
 ('store',)]

In [4]:
%%sql
SELECT 
    TABLE_NAME, COLUMN_NAME 
FROM 
    information_schema.KEY_COLUMN_USAGE 
WHERE 
    TABLE_SCHEMA = 'sakila' 
    AND CONSTRAINT_NAME = 'PRIMARY';


 * mysql+pymysql://root:***@localhost/sakila
19 rows affected.


[('actor', 'actor_id'),
 ('address', 'address_id'),
 ('category', 'category_id'),
 ('city', 'city_id'),
 ('country', 'country_id'),
 ('customer', 'customer_id'),
 ('employees', 'emp_id'),
 ('film', 'film_id'),
 ('film_actor', 'actor_id'),
 ('film_actor', 'film_id'),
 ('film_category', 'film_id'),
 ('film_category', 'category_id'),
 ('film_text', 'film_id'),
 ('inventory', 'inventory_id'),
 ('language', 'language_id'),
 ('payment', 'payment_id'),
 ('rental', 'rental_id'),
 ('staff', 'staff_id'),
 ('store', 'store_id')]

In [5]:
%%sql
SELECT 
    TABLE_NAME, COLUMN_NAME, CONSTRAINT_NAME, REFERENCED_TABLE_NAME, REFERENCED_COLUMN_NAME 
FROM 
    information_schema.KEY_COLUMN_USAGE 
WHERE 
    TABLE_SCHEMA = 'sakila' 
    AND REFERENCED_TABLE_NAME IS NOT NULL;


 * mysql+pymysql://root:***@localhost/sakila
22 rows affected.


[('address', 'city_id', 'fk_address_city', 'city', 'city_id'),
 ('city', 'country_id', 'fk_city_country', 'country', 'country_id'),
 ('customer', 'address_id', 'fk_customer_address', 'address', 'address_id'),
 ('customer', 'store_id', 'fk_customer_store', 'store', 'store_id'),
 ('film', 'language_id', 'fk_film_language', 'language', 'language_id'),
 ('film', 'original_language_id', 'fk_film_language_original', 'language', 'language_id'),
 ('film_actor', 'actor_id', 'fk_film_actor_actor', 'actor', 'actor_id'),
 ('film_actor', 'film_id', 'fk_film_actor_film', 'film', 'film_id'),
 ('film_category', 'category_id', 'fk_film_category_category', 'category', 'category_id'),
 ('film_category', 'film_id', 'fk_film_category_film', 'film', 'film_id'),
 ('inventory', 'film_id', 'fk_inventory_film', 'film', 'film_id'),
 ('inventory', 'store_id', 'fk_inventory_store', 'store', 'store_id'),
 ('payment', 'customer_id', 'fk_payment_customer', 'customer', 'customer_id'),
 ('payment', 'rental_id', 'fk_payment_rental', 'rental', 'rental_id'),
 ('payment', 'staff_id', 'fk_payment_staff', 'staff', 'staff_id'),
 ('rental', 'customer_id', 'fk_rental_customer', 'customer', 'customer_id'),
 ('rental', 'inventory_id', 'fk_rental_inventory', 'inventory', 'inventory_id'),
 ('rental', 'staff_id', 'fk_rental_staff', 'staff', 'staff_id'),
 ('staff', 'address_id', 'fk_staff_address', 'address', 'address_id'),
 ('staff', 'store_id', 'fk_staff_store', 'store', 'store_id'),
 ('store', 'address_id', 'fk_store_address', 'address', 'address_id'),
 ('store', 'manager_staff_id', 'fk_store_staff', 'staff', 'staff_id')]

In [6]:
%%sql
select * from actor limit 2

 * mysql+pymysql://root:***@localhost/sakila
2 rows affected.


[(1, 'PENELOPE', 'GUINESS', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (2, 'NICK', 'WAHLBERG', datetime.datetime(2006, 2, 15, 4, 34, 33))]

In [7]:
%%sql
select * from customer limit 2

 * mysql+pymysql://root:***@localhost/sakila
2 rows affected.


[(1, 1, 'MARY', 'SMITH', 'MARY.SMITH@sakilacustomer.org', 5, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (2, 1, 'PATRICIA', 'JOHNSON', 'PATRICIA.JOHNSON@sakilacustomer.org', 6, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20))]

In [8]:
%%sql
SELECT DISTINCT country FROM country limit 4;

 * mysql+pymysql://root:***@localhost/sakila
4 rows affected.


[('Afghanistan',), ('Algeria',), ('American Samoa',), ('Angola',)]

In [9]:
%%sql
SELECT * FROM customer WHERE active = 1 limit 5;

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(1, 1, 'MARY', 'SMITH', 'MARY.SMITH@sakilacustomer.org', 5, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (2, 1, 'PATRICIA', 'JOHNSON', 'PATRICIA.JOHNSON@sakilacustomer.org', 6, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (3, 1, 'LINDA', 'WILLIAMS', 'LINDA.WILLIAMS@sakilacustomer.org', 7, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (4, 2, 'BARBARA', 'JONES', 'BARBARA.JONES@sakilacustomer.org', 8, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (5, 1, 'ELIZABETH', 'BROWN', 'ELIZABETH.BROWN@sakilacustomer.org', 9, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20))]

In [10]:
%%sql
SELECT rental_id 
FROM rental 
WHERE customer_id = 1
limit 5

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(76,), (573,), (1185,), (1422,), (1476,)]

In [11]:
%%sql
SELECT * 
FROM film 
WHERE rental_duration > 5
limit 2

 * mysql+pymysql://root:***@localhost/sakila
2 rows affected.


[(1, 'ACADEMY DINOSAUR', 'A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies', 2006, 1, None, 6, Decimal('0.99'), 86, Decimal('20.99'), 'PG', 'Deleted Scenes,Behind the Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (3, 'ADAPTATION HOLES', 'A Astounding Reflection of a Lumberjack And a Car who must Sink a Lumberjack in A Baloon Factory', 2006, 1, None, 7, Decimal('2.99'), 50, Decimal('18.99'), 'NC-17', 'Trailers,Deleted Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42))]

In [12]:
%%sql
SELECT COUNT(*) AS total_films 
FROM film
WHERE replacement_cost > 15 AND replacement_cost < 20;


 * mysql+pymysql://root:***@localhost/sakila
1 rows affected.


[(214,)]

In [13]:
%%sql
SELECT COUNT(DISTINCT first_name) AS unique_first_names 
FROM actor;


 * mysql+pymysql://root:***@localhost/sakila
1 rows affected.


[(128,)]

In [14]:
%%sql
SELECT * 
FROM customer 
LIMIT 10;


 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[(1, 1, 'MARY', 'SMITH', 'MARY.SMITH@sakilacustomer.org', 5, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (2, 1, 'PATRICIA', 'JOHNSON', 'PATRICIA.JOHNSON@sakilacustomer.org', 6, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (3, 1, 'LINDA', 'WILLIAMS', 'LINDA.WILLIAMS@sakilacustomer.org', 7, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (4, 2, 'BARBARA', 'JONES', 'BARBARA.JONES@sakilacustomer.org', 8, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (5, 1, 'ELIZABETH', 'BROWN', 'ELIZABETH.BROWN@sakilacustomer.org', 9, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (6, 2, 'JENNIFER', 'DAVIS', 'JENNIFER.DAVIS@sakilacustomer.org', 10, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (7, 1, 'MARIA', 'MILLER', 'MARIA.MILLER@sakilacustomer.org', 11, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (8, 2, 'SUSAN', 'WILSON', 'SUSAN.WILSON@sakilacustomer.org', 12, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (9, 2, 'MARGARET', 'MOORE', 'MARGARET.MOORE@sakilacustomer.org', 13, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (10, 1, 'DOROTHY', 'TAYLOR', 'DOROTHY.TAYLOR@sakilacustomer.org', 14, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20))]

In [15]:
%%sql
-- 11. Display the first 3 records from the customer table whose first name starts with ‘b’.
SELECT * FROM customer WHERE first_name LIKE 'b%' LIMIT 3;

 * mysql+pymysql://root:***@localhost/sakila
3 rows affected.


[(4, 2, 'BARBARA', 'JONES', 'BARBARA.JONES@sakilacustomer.org', 8, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (14, 2, 'BETTY', 'WHITE', 'BETTY.WHITE@sakilacustomer.org', 18, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (31, 2, 'BRENDA', 'WRIGHT', 'BRENDA.WRIGHT@sakilacustomer.org', 35, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20))]

In [16]:
%%sql
-- 12. Display the names of the first 5 movies which are rated as ‘G’.
SELECT title FROM film WHERE rating = 'G' LIMIT 5;


 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[('ACE GOLDFINGER',),
 ('AFFAIR PREJUDICE',),
 ('AFRICAN EGG',),
 ('ALAMO VIDEOTAPE',),
 ('AMISTAD MIDSUMMER',)]

In [17]:
%%sql
-- 13. Find all customers whose first name starts with "a".
SELECT * FROM customer WHERE first_name LIKE 'a%';


 * mysql+pymysql://root:***@localhost/sakila
44 rows affected.


[(29, 2, 'ANGELA', 'HERNANDEZ', 'ANGELA.HERNANDEZ@sakilacustomer.org', 33, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (32, 1, 'AMY', 'LOPEZ', 'AMY.LOPEZ@sakilacustomer.org', 36, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (33, 2, 'ANNA', 'HILL', 'ANNA.HILL@sakilacustomer.org', 37, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (40, 2, 'AMANDA', 'CARTER', 'AMANDA.CARTER@sakilacustomer.org', 44, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (48, 1, 'ANN', 'EVANS', 'ANN.EVANS@sakilacustomer.org', 52, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (51, 1, 'ALICE', 'STEWART', 'ALICE.STEWART@sakilacustomer.org', 55, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (63, 1, 'ASHLEY', 'RICHARDSON', 'ASHLEY.RICHARDSON@sakilacustomer.org', 67, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (81, 1, 'ANDREA', 'HENDERSON', 'ANDREA.HENDERSON@sakilacustomer.org', 85, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (85, 2, 'ANNE', 'POWELL', 'ANNE.POWELL@sakilacustomer.org', 89, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (97, 2, 'ANNIE', 'RUSSELL', 'ANNIE.RUSSELL@sakilacustomer.org', 101, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (136, 2, 'ANITA', 'MORALES', 'ANITA.MORALES@sakilacustomer.org', 140, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (139, 1, 'AMBER', 'DIXON', 'AMBER.DIXON@sakilacustomer.org', 143, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (142, 1, 'APRIL', 'BURNS', 'APRIL.BURNS@sakilacustomer.org', 146, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (152, 1, 'ALICIA', 'MILLS', 'ALICIA.MILLS@sakilacustomer.org', 156, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (173, 1, 'AUDREY', 'RAY', 'AUDREY.RAY@sakilacustomer.org', 177, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (175, 1, 'ANNETTE', 'OLSON', 'ANNETTE.OLSON@sakilacustomer.org', 179, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (181, 2, 'ANA', 'BRADLEY', 'ANA.BRADLEY@sakilacustomer.org', 185, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (196, 1, 'ALMA', 'AUSTIN', 'ALMA.AUSTIN@sakilacustomer.org', 200, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (217, 2, 'AGNES', 'BISHOP', 'AGNES.BISHOP@sakilacustomer.org', 221, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (225, 1, 'ARLENE', 'HARVEY', 'ARLENE.HARVEY@sakilacustomer.org', 229, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (228, 2, 'ALLISON', 'STANLEY', 'ALLISON.STANLEY@sakilacustomer.org', 232, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (320, 2, 'ANTHONY', 'SCHWAB', 'ANTHONY.SCHWAB@sakilacustomer.org', 325, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (333, 2, 'ANDREW', 'PURDY', 'ANDREW.PURDY@sakilacustomer.org', 338, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (346, 1, 'ARTHUR', 'SIMPKINS', 'ARTHUR.SIMPKINS@sakilacustomer.org', 351, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (352, 1, 'ALBERT', 'CROUSE', 'ALBERT.CROUSE@sakilacustomer.org', 357, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (367, 1, 'ADAM', 'G

In [18]:
%%sql
-- 14. Find all customers whose first name ends with "a".
SELECT * FROM customer WHERE first_name LIKE '%a';


 * mysql+pymysql://root:***@localhost/sakila
96 rows affected.


[(2, 1, 'PATRICIA', 'JOHNSON', 'PATRICIA.JOHNSON@sakilacustomer.org', 6, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (3, 1, 'LINDA', 'WILLIAMS', 'LINDA.WILLIAMS@sakilacustomer.org', 7, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (4, 2, 'BARBARA', 'JONES', 'BARBARA.JONES@sakilacustomer.org', 8, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (7, 1, 'MARIA', 'MILLER', 'MARIA.MILLER@sakilacustomer.org', 11, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (11, 2, 'LISA', 'ANDERSON', 'LISA.ANDERSON@sakilacustomer.org', 15, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (16, 2, 'SANDRA', 'MARTIN', 'SANDRA.MARTIN@sakilacustomer.org', 20, 0, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (17, 1, 'DONNA', 'THOMPSON', 'DONNA.THOMPSON@sakilacustomer.org', 21, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (22, 1, 'LAURA', 'RODRIGUEZ', 'LAURA.RODRIGUEZ@sakilacustomer.org', 26, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (26, 2, 'JESSICA', 'HALL', 'JESSICA.HALL@sakilacustomer.org', 30, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (28, 1, 'CYNTHIA', 'YOUNG', 'CYNTHIA.YOUNG@sakilacustomer.org', 32, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (29, 2, 'ANGELA', 'HERNANDEZ', 'ANGELA.HERNANDEZ@sakilacustomer.org', 33, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (30, 1, 'MELISSA', 'KING', 'MELISSA.KING@sakilacustomer.org', 34, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (31, 2, 'BRENDA', 'WRIGHT', 'BRENDA.WRIGHT@sakilacustomer.org', 35, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (33, 2, 'ANNA', 'HILL', 'ANNA.HILL@sakilacustomer.org', 37, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (34, 2, 'REBECCA', 'SCOTT', 'REBECCA.SCOTT@sakilacustomer.org', 38, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (35, 2, 'VIRGINIA', 'GREEN', 'VIRGINIA.GREEN@sakilacustomer.org', 39, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (37, 1, 'PAMELA', 'BAKER', 'PAMELA.BAKER@sakilacustomer.org', 41, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (38, 1, 'MARTHA', 'GONZALEZ', 'MARTHA.GONZALEZ@sakilacustomer.org', 42, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (39, 1, 'DEBRA', 'NELSON', 'DEBRA.NELSON@sakilacustomer.org', 43, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (40, 2, 'AMANDA', 'CARTER', 'AMANDA.CARTER@sakilacustomer.org', 44, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (54, 1, 'TERESA', 'ROGERS', 'TERESA.ROGERS@sakilacustomer.org', 58, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (56, 1, 'GLORIA', 'COOK', 'GLORIA.COOK@sakilacustomer.org', 60, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (70, 2, 'CHRISTINA', 'RAMIREZ', 'CHRISTINA.RAMIREZ@sakilacustomer.org', 74, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (72, 2, 'THERESA', 'WATSON', 'THERESA.WATSON@sakilacustomer.org', 76, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (81, 1, 'ANDREA', 'HENDERSON', 'ANDREA.HENDERSON@sakilacustomer.org', 85, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (84, 2, 'SA

In [19]:
%%sql
-- 15. Display the list of first 4 cities which start and end with ‘a’.
SELECT city FROM city WHERE city LIKE 'a%a' LIMIT 4;


 * mysql+pymysql://root:***@localhost/sakila
4 rows affected.


[('Abha',), ('Acuña',), ('Adana',), ('Addis Abeba',)]

In [20]:
%%sql
-- 16. Find all customers whose first name has "NI" in any position.
SELECT * FROM customer WHERE first_name LIKE '%NI%';


 * mysql+pymysql://root:***@localhost/sakila
29 rows affected.


[(6, 2, 'JENNIFER', 'DAVIS', 'JENNIFER.DAVIS@sakilacustomer.org', 10, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (35, 2, 'VIRGINIA', 'GREEN', 'VIRGINIA.GREEN@sakilacustomer.org', 39, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (41, 1, 'STEPHANIE', 'MITCHELL', 'STEPHANIE.MITCHELL@sakilacustomer.org', 45, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (66, 2, 'JANICE', 'WARD', 'JANICE.WARD@sakilacustomer.org', 70, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (68, 1, 'NICOLE', 'PETERSON', 'NICOLE.PETERSON@sakilacustomer.org', 72, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (74, 1, 'DENISE', 'KELLY', 'DENISE.KELLY@sakilacustomer.org', 78, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (88, 2, 'BONNIE', 'HUGHES', 'BONNIE.HUGHES@sakilacustomer.org', 92, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (97, 2, 'ANNIE', 'RUSSELL', 'ANNIE.RUSSELL@sakilacustomer.org', 101, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (106, 1, 'CONNIE', 'WALLACE', 'CONNIE.WALLACE@sakilacustomer.org', 110, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (131, 2, 'MONICA', 'HICKS', 'MONICA.HICKS@sakilacustomer.org', 135, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (135, 2, 'JUANITA', 'MASON', 'JUANITA.MASON@sakilacustomer.org', 139, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (136, 2, 'ANITA', 'MORALES', 'ANITA.MORALES@sakilacustomer.org', 140, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (150, 2, 'DANIELLE', 'DANIELS', 'DANIELLE.DANIELS@sakilacustomer.org', 154, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (158, 1, 'VERONICA', 'STONE', 'VERONICA.STONE@sakilacustomer.org', 162, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (172, 1, 'BERNICE', 'WILLIS', 'BERNICE.WILLIS@sakilacustomer.org', 176, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (188, 1, 'MELANIE', 'ARMSTRONG', 'MELANIE.ARMSTRONG@sakilacustomer.org', 192, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (239, 2, 'MINNIE', 'ROMERO', 'MINNIE.ROMERO@sakilacustomer.org', 243, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (265, 2, 'JENNIE', 'TERRY', 'JENNIE.TERRY@sakilacustomer.org', 270, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (268, 1, 'NINA', 'SOTO', 'NINA.SOTO@sakilacustomer.org', 273, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (284, 1, 'SONIA', 'GREGORY', 'SONIA.GREGORY@sakilacustomer.org', 289, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (291, 1, 'TONI', 'HOLT', 'TONI.HOLT@sakilacustomer.org', 296, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (310, 2, 'DANIEL', 'CABRAL', 'DANIEL.CABRAL@sakilacustomer.org', 315, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (338, 1, 'DENNIS', 'GILMAN', 'DENNIS.GILMAN@sakilacustomer.org', 343, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (362, 1, 'NICHOLAS', 'BARFIELD', 'NICHOLAS.BARFIELD@sakilacustomer.org', 367, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (398, 1, 'ANTONIO', 'MEEK', 'ANTONIO.MEEK@sakilacustomer.org', 403, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime

In [21]:
%%sql
-- 17. Find all customers whose first name has "r" in the second position.
SELECT * FROM customer WHERE first_name LIKE '_r%';


 * mysql+pymysql://root:***@localhost/sakila
45 rows affected.


[(31, 2, 'BRENDA', 'WRIGHT', 'BRENDA.WRIGHT@sakilacustomer.org', 35, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (47, 1, 'FRANCES', 'PARKER', 'FRANCES.PARKER@sakilacustomer.org', 51, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (76, 2, 'IRENE', 'PRICE', 'IRENE.PRICE@sakilacustomer.org', 80, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (102, 1, 'CRYSTAL', 'FORD', 'CRYSTAL.FORD@sakilacustomer.org', 106, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (108, 1, 'TRACY', 'COLE', 'TRACY.COLE@sakilacustomer.org', 112, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (114, 2, 'GRACE', 'ELLIS', 'GRACE.ELLIS@sakilacustomer.org', 118, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (160, 2, 'ERIN', 'DUNN', 'ERIN.DUNN@sakilacustomer.org', 164, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (169, 2, 'ERICA', 'MATTHEWS', 'ERICA.MATTHEWS@sakilacustomer.org', 173, 0, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (187, 2, 'BRITTANY', 'RILEY', 'BRITTANY.RILEY@sakilacustomer.org', 191, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (194, 2, 'KRISTEN', 'CHAVEZ', 'KRISTEN.CHAVEZ@sakilacustomer.org', 198, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (214, 1, 'KRISTIN', 'JOHNSTON', 'KRISTIN.JOHNSTON@sakilacustomer.org', 218, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (225, 1, 'ARLENE', 'HARVEY', 'ARLENE.HARVEY@sakilacustomer.org', 229, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (255, 2, 'IRMA', 'PEARSON', 'IRMA.PEARSON@sakilacustomer.org', 260, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (273, 2, 'PRISCILLA', 'LOWE', 'PRISCILLA.LOWE@sakilacustomer.org', 278, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (276, 1, 'BRANDY', 'GRAVES', 'BRANDY.GRAVES@sakilacustomer.org', 281, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (280, 2, 'TRACEY', 'BARRETT', 'TRACEY.BARRETT@sakilacustomer.org', 285, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (290, 1, 'KRISTINA', 'CHAMBERS', 'KRISTINA.CHAMBERS@sakilacustomer.org', 295, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (298, 1, 'ERIKA', 'PENA', 'ERIKA.PENA@sakilacustomer.org', 303, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (318, 1, 'BRIAN', 'WYMAN', 'BRIAN.WYMAN@sakilacustomer.org', 323, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (329, 2, 'FRANK', 'WAGGONER', 'FRANK.WAGGONER@sakilacustomer.org', 334, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (331, 1, 'ERIC', 'ROBERT', 'ERIC.ROBERT@sakilacustomer.org', 336, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (335, 1, 'GREGORY', 'MAULDIN', 'GREGORY.MAULDIN@sakilacustomer.org', 340, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (346, 1, 'ARTHUR', 'SIMPKINS', 'ARTHUR.SIMPKINS@sakilacustomer.org', 351, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (365, 2, 'BRUCE', 'SCHWARZ', 'BRUCE.SCHWARZ@sakilacustomer.org', 370, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (366, 1, 'BRANDON', 'HUEY', 'BRANDON.HUEY@sakilacustomer.org', 371, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 

In [22]:
%%sql
-- 18. Find all customers whose first name starts with "a" and is at least 5 characters in length.
SELECT * FROM customer WHERE first_name LIKE 'a%' AND LENGTH(first_name) >= 5;


 * mysql+pymysql://root:***@localhost/sakila
34 rows affected.


[(29, 2, 'ANGELA', 'HERNANDEZ', 'ANGELA.HERNANDEZ@sakilacustomer.org', 33, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (40, 2, 'AMANDA', 'CARTER', 'AMANDA.CARTER@sakilacustomer.org', 44, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (51, 1, 'ALICE', 'STEWART', 'ALICE.STEWART@sakilacustomer.org', 55, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (63, 1, 'ASHLEY', 'RICHARDSON', 'ASHLEY.RICHARDSON@sakilacustomer.org', 67, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (81, 1, 'ANDREA', 'HENDERSON', 'ANDREA.HENDERSON@sakilacustomer.org', 85, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (97, 2, 'ANNIE', 'RUSSELL', 'ANNIE.RUSSELL@sakilacustomer.org', 101, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (136, 2, 'ANITA', 'MORALES', 'ANITA.MORALES@sakilacustomer.org', 140, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (139, 1, 'AMBER', 'DIXON', 'AMBER.DIXON@sakilacustomer.org', 143, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (142, 1, 'APRIL', 'BURNS', 'APRIL.BURNS@sakilacustomer.org', 146, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (152, 1, 'ALICIA', 'MILLS', 'ALICIA.MILLS@sakilacustomer.org', 156, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (173, 1, 'AUDREY', 'RAY', 'AUDREY.RAY@sakilacustomer.org', 177, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (175, 1, 'ANNETTE', 'OLSON', 'ANNETTE.OLSON@sakilacustomer.org', 179, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (217, 2, 'AGNES', 'BISHOP', 'AGNES.BISHOP@sakilacustomer.org', 221, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (225, 1, 'ARLENE', 'HARVEY', 'ARLENE.HARVEY@sakilacustomer.org', 229, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (228, 2, 'ALLISON', 'STANLEY', 'ALLISON.STANLEY@sakilacustomer.org', 232, 1, datetime.datetime(2006, 2, 14, 22, 4, 36), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (320, 2, 'ANTHONY', 'SCHWAB', 'ANTHONY.SCHWAB@sakilacustomer.org', 325, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (333, 2, 'ANDREW', 'PURDY', 'ANDREW.PURDY@sakilacustomer.org', 338, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (346, 1, 'ARTHUR', 'SIMPKINS', 'ARTHUR.SIMPKINS@sakilacustomer.org', 351, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (352, 1, 'ALBERT', 'CROUSE', 'ALBERT.CROUSE@sakilacustomer.org', 357, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (375, 2, 'AARON', 'SELBY', 'AARON.SELBY@sakilacustomer.org', 380, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (398, 1, 'ANTONIO', 'MEEK', 'ANTONIO.MEEK@sakilacustomer.org', 403, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (412, 2, 'ALLEN', 'BUTTERFIELD', 'ALLEN.BUTTERFIELD@sakilacustomer.org', 417, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (423, 2, 'ALFRED', 'CASILLAS', 'ALFRED.CASILLAS@sakilacustomer.org', 428, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (439, 2, 'ALEXANDER', 'FENNELL', 'ALEXANDER.FENNELL@sakilacustomer.org', 444, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (467, 2, 'ALVIN', 'DELOACH', 'ALVIN.DELOACH@sakilacustomer.org', 472, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), dateti

In [23]:
%%sql
-- 19. Find all customers whose first name starts with "a" and ends with "o".
SELECT * FROM customer WHERE first_name LIKE 'a%o';


 * mysql+pymysql://root:***@localhost/sakila
4 rows affected.


[(398, 1, 'ANTONIO', 'MEEK', 'ANTONIO.MEEK@sakilacustomer.org', 403, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (556, 2, 'ARMANDO', 'GRUBER', 'ARMANDO.GRUBER@sakilacustomer.org', 562, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (567, 2, 'ALFREDO', 'MCADAMS', 'ALFREDO.MCADAMS@sakilacustomer.org', 573, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20)),
 (568, 2, 'ALBERTO', 'HENNING', 'ALBERTO.HENNING@sakilacustomer.org', 574, 1, datetime.datetime(2006, 2, 14, 22, 4, 37), datetime.datetime(2006, 2, 15, 4, 57, 20))]

In [24]:
%%sql
-- 20. Get the films with PG and PG-13 ratings using the IN operator.
SELECT * FROM film WHERE rating IN ('PG', 'PG-13') limit 5;


 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(1, 'ACADEMY DINOSAUR', 'A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies', 2006, 1, None, 6, Decimal('0.99'), 86, Decimal('20.99'), 'PG', 'Deleted Scenes,Behind the Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (6, 'AGENT TRUMAN', 'A Intrepid Panorama of a Robot And a Boy who must Escape a Sumo Wrestler in Ancient China', 2006, 1, None, 3, Decimal('2.99'), 169, Decimal('17.99'), 'PG', 'Deleted Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (7, 'AIRPLANE SIERRA', 'A Touching Saga of a Hunter And a Butler who must Discover a Butler in A Jet Boat', 2006, 1, None, 6, Decimal('4.99'), 62, Decimal('28.99'), 'PG-13', 'Trailers,Deleted Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (9, 'ALABAMA DEVIL', 'A Thoughtful Panorama of a Database Administrator And a Mad Scientist who must Outgun a Mad Scientist in A Jet Boat', 2006, 1, None, 3, Decimal('2.99'), 114, Decimal('21.99'), 'PG-13', 'Trailers,Deleted Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (12, 'ALASKA PHANTOM', 'A Fanciful Saga of a Hunter And a Pastry Chef who must Vanquish a Boy in Australia', 2006, 1, None, 6, Decimal('0.99'), 136, Decimal('22.99'), 'PG', 'Commentaries,Deleted Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42))]

In [25]:
%%sql
-- 21. Get the films with length between 50 and 100 using the BETWEEN operator.
SELECT * FROM film WHERE length BETWEEN 50 AND 100 limit 5;


 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(1, 'ACADEMY DINOSAUR', 'A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies', 2006, 1, None, 6, Decimal('0.99'), 86, Decimal('20.99'), 'PG', 'Deleted Scenes,Behind the Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (3, 'ADAPTATION HOLES', 'A Astounding Reflection of a Lumberjack And a Car who must Sink a Lumberjack in A Baloon Factory', 2006, 1, None, 7, Decimal('2.99'), 50, Decimal('18.99'), 'NC-17', 'Trailers,Deleted Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (7, 'AIRPLANE SIERRA', 'A Touching Saga of a Hunter And a Butler who must Discover a Butler in A Jet Boat', 2006, 1, None, 6, Decimal('4.99'), 62, Decimal('28.99'), 'PG-13', 'Trailers,Deleted Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (8, 'AIRPORT POLLOCK', 'A Epic Tale of a Moose And a Girl who must Confront a Monkey in Ancient India', 2006, 1, None, 6, Decimal('4.99'), 54, Decimal('15.99'), 'R', 'Trailers', datetime.datetime(2006, 2, 15, 5, 3, 42)),
 (10, 'ALADDIN CALENDAR', 'A Action-Packed Tale of a Man And a Lumberjack who must Reach a Feminist in Ancient China', 2006, 1, None, 6, Decimal('4.99'), 63, Decimal('24.99'), 'NC-17', 'Trailers,Deleted Scenes', datetime.datetime(2006, 2, 15, 5, 3, 42))]

In [26]:
%%sql
-- 22. Get the top 50 actors using the LIMIT operator.
SELECT * FROM actor LIMIT 50;


 * mysql+pymysql://root:***@localhost/sakila
50 rows affected.


[(1, 'PENELOPE', 'GUINESS', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (2, 'NICK', 'WAHLBERG', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (3, 'ED', 'CHASE', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (4, 'JENNIFER', 'DAVIS', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (5, 'JOHNNY', 'LOLLOBRIGIDA', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (6, 'BETTE', 'NICHOLSON', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (7, 'GRACE', 'MOSTEL', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (8, 'MATTHEW', 'JOHANSSON', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (9, 'JOE', 'SWANK', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (10, 'CHRISTIAN', 'GABLE', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (11, 'ZERO', 'CAGE', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (12, 'KARL', 'BERRY', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (13, 'UMA', 'WOOD', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (14, 'VIVIEN', 'BERGEN', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (15, 'CUBA', 'OLIVIER', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (16, 'FRED', 'COSTNER', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (17, 'HELEN', 'VOIGHT', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (18, 'DAN', 'TORN', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (19, 'BOB', 'FAWCETT', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (20, 'LUCILLE', 'TRACY', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (21, 'KIRSTEN', 'PALTROW', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (22, 'ELVIS', 'MARX', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (23, 'SANDRA', 'KILMER', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (24, 'CAMERON', 'STREEP', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (25, 'KEVIN', 'BLOOM', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (26, 'RIP', 'CRAWFORD', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (27, 'JULIA', 'MCQUEEN', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (28, 'WOODY', 'HOFFMAN', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (29, 'ALEC', 'WAYNE', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (30, 'SANDRA', 'PECK', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (31, 'SISSY', 'SOBIESKI', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (32, 'TIM', 'HACKMAN', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (33, 'MILLA', 'PECK', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (34, 'AUDREY', 'OLIVIER', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (35, 'JUDY', 'DEAN', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (36, 'BURT', 'DUKAKIS', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (37, 'VAL', 'BOLGER', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (38, 'TOM', 'MCKELLEN', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (39, 'GOLDIE', 'BRODY', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (40, 'JOHNNY', 'CAGE', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (41, 'JODIE', 'DEGENERES', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (42, 'TOM', 'MIRANDA', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (43, 'KIRK', 'JOVOVICH', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (44, 'NICK', 'STALLONE', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (45, 'REESE', 'KILMER', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (46, 'PARKER', 'GOLDBERG', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (47, 'JULIA', 'BARRYMORE', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (48, 'FRANCES', 'DAY-LEWIS', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (49, 'ANNE', 'CRONYN', datetime.datetime(2006, 2, 15, 4, 34, 33)),
 (50, 'NATALIE', 'HOPKINS', datetime.datetime(2006, 2, 15, 4, 34, 33))]

In [27]:
%%sql
-- 23. Get the distinct film IDs from the inventory table.
SELECT DISTINCT film_id FROM inventory limit 5;


 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(1,), (2,), (3,), (4,), (5,)]

# Functions

In [28]:
%%sql
-- Question 1: Retrieve the total number of rentals made in the Sakila database.
    
SELECT COUNT(*) AS total_rentals FROM rental;

 * mysql+pymysql://root:***@localhost/sakila
1 rows affected.


[(16044,)]

In [29]:
%%sql
-- Question 2: Find the average rental duration (in days) of movies rented from the Sakila database.

SELECT AVG(rental_duration) AS average_rental_duration FROM film;

 * mysql+pymysql://root:***@localhost/sakila
1 rows affected.


[(Decimal('4.9850'),)]

In [30]:
%%sql
-- Question 3: Display the first name and last name of customers in uppercase.
    
SELECT UPPER(first_name) AS first_name, UPPER(last_name) AS last_name FROM customer limit 5;

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[('MARY', 'SMITH'),
 ('PATRICIA', 'JOHNSON'),
 ('LINDA', 'WILLIAMS'),
 ('BARBARA', 'JONES'),
 ('ELIZABETH', 'BROWN')]

In [31]:
%%sql
-- Question 4: Extract the month from the rental date and display it alongside the rental ID.

SELECT rental_id, MONTH(rental_date) AS rental_month FROM rental  limit 5;

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(1, 5), (2, 5), (3, 5), (4, 5), (5, 5)]

In [32]:
%%sql
-- Question 5: Retrieve the count of rentals for each customer (display customer ID and the count of rentals).

SELECT customer_id, COUNT(*) AS rental_count 
FROM rental 
GROUP BY customer_id limit 5;

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(1, 32), (2, 27), (3, 26), (4, 22), (5, 38)]

In [33]:
%%sql
select * from payment limit 3;

 * mysql+pymysql://root:***@localhost/sakila
3 rows affected.


[(1, 1, 1, 76, Decimal('2.99'), datetime.datetime(2005, 5, 25, 11, 30, 37), datetime.datetime(2006, 2, 15, 22, 12, 30)),
 (2, 1, 1, 573, Decimal('0.99'), datetime.datetime(2005, 5, 28, 10, 35, 23), datetime.datetime(2006, 2, 15, 22, 12, 30)),
 (3, 1, 1, 1185, Decimal('5.99'), datetime.datetime(2005, 6, 15, 0, 54, 12), datetime.datetime(2006, 2, 15, 22, 12, 30))]

In [34]:
%%sql
-- Question 6: Find the total revenue generated by each store.

SELECT store.store_id, SUM(payment.amount) AS total_revenue 
FROM payment
JOIN staff ON payment.staff_id = staff.staff_id
JOIN store ON staff.store_id = store.store_id
GROUP BY store.store_id;


 * mysql+pymysql://root:***@localhost/sakila
2 rows affected.


[(1, Decimal('33482.50')), (2, Decimal('33924.06'))]

In [35]:
%%sql
-- Question 7: Determine the total number of rentals for each category of movies.

SELECT category.name AS category, COUNT(*) AS rental_count 
FROM rental
JOIN inventory ON rental.inventory_id = inventory.inventory_id
JOIN film ON inventory.film_id = film.film_id
JOIN film_category ON film.film_id = film_category.film_id
JOIN category ON film_category.category_id = category.category_id
GROUP BY category.name;

 * mysql+pymysql://root:***@localhost/sakila
16 rows affected.


[('Action', 1112),
 ('Animation', 1166),
 ('Children', 945),
 ('Classics', 939),
 ('Comedy', 941),
 ('Documentary', 1050),
 ('Drama', 1060),
 ('Family', 1096),
 ('Foreign', 1033),
 ('Games', 969),
 ('Horror', 846),
 ('Music', 830),
 ('New', 940),
 ('Sci-Fi', 1101),
 ('Sports', 1179),
 ('Travel', 837)]

In [36]:
%%sql
-- Question 8: Find the average rental rate of movies in each language.

SELECT language.name AS language, AVG(film.rental_rate) AS average_rental_rate 
FROM film
JOIN language ON film.language_id = language.language_id
GROUP BY language.name;

 * mysql+pymysql://root:***@localhost/sakila
1 rows affected.


[('English', Decimal('2.980000'))]

# Joins

In [37]:
%%sql
-- Question 9: Display the title of the movie, customer’s first name, and last name who rented it.

SELECT film.title, customer.first_name, customer.last_name 
FROM rental
JOIN inventory ON rental.inventory_id = inventory.inventory_id
JOIN film ON inventory.film_id = film.film_id
JOIN customer ON rental.customer_id = customer.customer_id limit 5;

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[('ACADEMY DINOSAUR', 'JOEL', 'FRANCISCO'),
 ('ACADEMY DINOSAUR', 'GABRIEL', 'HARDER'),
 ('ACADEMY DINOSAUR', 'DIANNE', 'SHELTON'),
 ('ACADEMY DINOSAUR', 'NORMAN', 'CURRIER'),
 ('ACADEMY DINOSAUR', 'BEATRICE', 'ARNOLD')]

In [38]:
%%sql
-- Question 10: Retrieve the names of all actors who have appeared in the film "GONE TROUBLE."

SELECT actor.first_name, actor.last_name 
FROM actor
JOIN film_actor ON actor.actor_id = film_actor.actor_id
JOIN film ON film_actor.film_id = film.film_id
WHERE film.title = 'GONE TROUBLE';

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[('SISSY', 'SOBIESKI'),
 ('RICHARD', 'PENN'),
 ('EMILY', 'DEE'),
 ('CHRIS', 'DEPP'),
 ('GROUCHO', 'WILLIAMS')]

In [39]:
%%sql
-- Question 11: Retrieve the customer names along with the total amount they've spent on rentals.

SELECT customer.first_name, customer.last_name, SUM(payment.amount) AS total_spent
FROM customer
JOIN payment ON customer.customer_id = payment.customer_id
GROUP BY customer.customer_id limit 7;

 * mysql+pymysql://root:***@localhost/sakila
7 rows affected.


[('MARY', 'SMITH', Decimal('118.68')),
 ('PATRICIA', 'JOHNSON', Decimal('128.73')),
 ('LINDA', 'WILLIAMS', Decimal('135.74')),
 ('BARBARA', 'JONES', Decimal('81.78')),
 ('ELIZABETH', 'BROWN', Decimal('144.62')),
 ('JENNIFER', 'DAVIS', Decimal('93.72')),
 ('MARIA', 'MILLER', Decimal('151.67'))]

In [40]:
%%sql
-- Question 12: List the titles of movies rented by each customer in a particular city (e.g., 'London').

SELECT customer.first_name, customer.last_name, film.title
FROM customer
JOIN address ON customer.address_id = address.address_id
JOIN city ON address.city_id = city.city_id
JOIN rental ON customer.customer_id = rental.customer_id
JOIN inventory ON rental.inventory_id = inventory.inventory_id
JOIN film ON inventory.film_id = film.film_id
WHERE city.city = 'London'
GROUP BY customer.customer_id, film.title;

 * mysql+pymysql://root:***@localhost/sakila
47 rows affected.


[('MATTIE', 'HOFFMAN', 'CONQUERER NUTS'),
 ('MATTIE', 'HOFFMAN', 'WRATH MILE'),
 ('MATTIE', 'HOFFMAN', 'COLDBLOODED DARLING'),
 ('MATTIE', 'HOFFMAN', 'DARKNESS WAR'),
 ('MATTIE', 'HOFFMAN', 'WATERSHIP FRONTIER'),
 ('MATTIE', 'HOFFMAN', 'JAWBREAKER BROOKLYN'),
 ('MATTIE', 'HOFFMAN', 'TIGHTS DAWN'),
 ('MATTIE', 'HOFFMAN', 'CHICKEN HELLFIGHTERS'),
 ('MATTIE', 'HOFFMAN', 'DOOM DANCING'),
 ('MATTIE', 'HOFFMAN', 'PITY BOUND'),
 ('MATTIE', 'HOFFMAN', 'EMPIRE MALKOVICH'),
 ('MATTIE', 'HOFFMAN', 'ACADEMY DINOSAUR'),
 ('MATTIE', 'HOFFMAN', 'FACTORY DRAGON'),
 ('MATTIE', 'HOFFMAN', 'VELVET TERMINATOR'),
 ('MATTIE', 'HOFFMAN', 'TRACY CIDER'),
 ('MATTIE', 'HOFFMAN', 'FROGMEN BREAKING'),
 ('MATTIE', 'HOFFMAN', 'JUMPING WRATH'),
 ('MATTIE', 'HOFFMAN', 'WOMEN DORADO'),
 ('MATTIE', 'HOFFMAN', 'VIDEOTAPE ARSENIC'),
 ('MATTIE', 'HOFFMAN', 'DRIFTER COMMANDMENTS'),
 ('MATTIE', 'HOFFMAN', 'FLATLINERS KILLER'),
 ('MATTIE', 'HOFFMAN', 'PRINCESS GIANT'),
 ('CECIL', 'VINES', 'MOONSHINE CABIN'),
 ('CECIL', 'VINES', 'DAISY MENAGERIE'),
 ('CECIL', 'VINES', 'NONE SPIKING'),
 ('CECIL', 'VINES', 'RANGE MOONWALKER'),
 ('CECIL', 'VINES', 'TOMATOES HELLFIGHTERS'),
 ('CECIL', 'VINES', 'BLOOD ARGONAUTS'),
 ('CECIL', 'VINES', 'DARKNESS WAR'),
 ('CECIL', 'VINES', 'TELEMARK HEARTBREAKERS'),
 ('CECIL', 'VINES', 'SHOOTIST SUPERFLY'),
 ('CECIL', 'VINES', 'PURPLE MOVIE'),
 ('CECIL', 'VINES', 'EDGE KISSING'),
 ('CECIL', 'VINES', 'CHARIOTS CONSPIRACY'),
 ('CECIL', 'VINES', 'CAT CONEHEADS'),
 ('CECIL', 'VINES', 'ARABIA DOGMA'),
 ('CECIL', 'VINES', 'BACKLASH UNDEFEATED'),
 ('CECIL', 'VINES', 'AMADEUS HOLY'),
 ('CECIL', 'VINES', 'JEEPERS WEDDING'),
 ('CECIL', 'VINES', 'COMA HEAD'),
 ('CECIL', 'VINES', 'KANE EXORCIST'),
 ('CECIL', 'VINES', 'INVASION CYCLONE'),
 ('CECIL', 'VINES', 'DISTURBING SCARFACE'),
 ('CECIL', 'VINES', 'GRIT CLOCKWORK'),
 ('CECIL', 'VINES', 'LABYRINTH LEAGUE'),
 ('CECIL', 'VINES', 'TRIP NEWTON'),
 ('CECIL', 'VINES', 'RAGE GAMES')]

# Advanced Joins and GROUP BY

In [41]:
%%sql
-- Question 13: Display the top 5 rented movies along with the number of times they've been rented.

SELECT film.title, COUNT(rental.rental_id) AS rental_count
FROM rental
JOIN inventory ON rental.inventory_id = inventory.inventory_id
JOIN film ON inventory.film_id = film.film_id
GROUP BY film.title
ORDER BY rental_count DESC
LIMIT 5;

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[('BUCKET BROTHERHOOD', 34),
 ('ROCKETEER MOTHER', 33),
 ('FORWARD TEMPLE', 32),
 ('GRIT CLOCKWORK', 32),
 ('JUGGLER HARDLY', 32)]

In [42]:
%%sql
-- Question 14: Determine the customers who have rented movies from both stores (store ID 1 and store ID 2).

SELECT customer.customer_id, customer.first_name, customer.last_name
FROM customer
JOIN rental ON customer.customer_id = rental.customer_id
JOIN inventory ON rental.inventory_id = inventory.inventory_id
WHERE inventory.store_id IN (1, 2)
GROUP BY customer.customer_id, customer.first_name, customer.last_name
HAVING COUNT(DISTINCT inventory.store_id) = 2 LIMIT 7;

 * mysql+pymysql://root:***@localhost/sakila
7 rows affected.


[(1, 'MARY', 'SMITH'),
 (2, 'PATRICIA', 'JOHNSON'),
 (3, 'LINDA', 'WILLIAMS'),
 (4, 'BARBARA', 'JONES'),
 (5, 'ELIZABETH', 'BROWN'),
 (6, 'JENNIFER', 'DAVIS'),
 (7, 'MARIA', 'MILLER')]

# Windows Function

In [43]:
%%sql
-- 1. Rank the customers based on the total amount they've spent on rentals.

SELECT c.customer_id, c.first_name, c.last_name, SUM(amount) AS total_spent,
       RANK() OVER (ORDER BY SUM(amount) DESC) AS spending_rank
FROM customer as c
JOIN payment ON c.customer_id = payment.customer_id
GROUP BY customer_id, first_name, last_name limit 10;

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[(526, 'KARL', 'SEAL', Decimal('221.55'), 1),
 (148, 'ELEANOR', 'HUNT', Decimal('216.54'), 2),
 (144, 'CLARA', 'SHAW', Decimal('195.58'), 3),
 (137, 'RHONDA', 'KENNEDY', Decimal('194.61'), 4),
 (178, 'MARION', 'SNYDER', Decimal('194.61'), 4),
 (459, 'TOMMY', 'COLLAZO', Decimal('186.62'), 6),
 (469, 'WESLEY', 'BULL', Decimal('177.60'), 7),
 (468, 'TIM', 'CARY', Decimal('175.61'), 8),
 (236, 'MARCIA', 'DEAN', Decimal('175.58'), 9),
 (181, 'ANA', 'BRADLEY', Decimal('174.66'), 10)]

In [44]:
%%sql
-- 2. Calculate the cumulative revenue generated by each film over time.

SELECT film.title, payment.payment_date, 
       SUM(payment.amount) OVER (PARTITION BY film.film_id ORDER BY payment.payment_date) AS cumulative_revenue
FROM payment
JOIN rental ON payment.rental_id = rental.rental_id
JOIN inventory ON rental.inventory_id = inventory.inventory_id
JOIN film ON inventory.film_id = film.film_id LIMIT 7;

 * mysql+pymysql://root:***@localhost/sakila
7 rows affected.


[('ACADEMY DINOSAUR', datetime.datetime(2005, 5, 27, 7, 3, 28), Decimal('0.99')),
 ('ACADEMY DINOSAUR', datetime.datetime(2005, 5, 30, 20, 21, 7), Decimal('2.98')),
 ('ACADEMY DINOSAUR', datetime.datetime(2005, 6, 15, 2, 57, 51), Decimal('3.97')),
 ('ACADEMY DINOSAUR', datetime.datetime(2005, 6, 17, 20, 24), Decimal('4.96')),
 ('ACADEMY DINOSAUR', datetime.datetime(2005, 6, 21, 0, 30, 26), Decimal('6.95')),
 ('ACADEMY DINOSAUR', datetime.datetime(2005, 7, 7, 10, 41, 31), Decimal('7.94')),
 ('ACADEMY DINOSAUR', datetime.datetime(2005, 7, 7, 20, 59, 6), Decimal('8.93'))]

In [45]:
%%sql
-- 3. Determine the average rental duration for each film, considering films with similar lengths.
    
SELECT film.length, film.title, AVG(rental.return_date - rental.rental_date) AS average_rental_duration
FROM film
JOIN inventory ON film.film_id = inventory.film_id
JOIN rental ON inventory.inventory_id = rental.inventory_id
GROUP BY film.length, film.title LIMIT 7;

 * mysql+pymysql://root:***@localhost/sakila
7 rows affected.


[(86, 'ACADEMY DINOSAUR', Decimal('23886859.0909')),
 (48, 'ACE GOLDFINGER', Decimal('17161883.3333')),
 (50, 'ADAPTATION HOLES', Decimal('9178383.3333')),
 (117, 'AFFAIR PREJUDICE', Decimal('23552472.7273')),
 (130, 'AFRICAN EGG', Decimal('25913172.7273')),
 (169, 'AGENT TRUMAN', Decimal('15755133.3333')),
 (62, 'AIRPLANE SIERRA', Decimal('9144873.3333'))]

In [46]:
%%sql
-- 4. Identify the top 3 films in each category based on their rental counts.

SELECT category, title, rental_count
FROM (
    SELECT category.name AS category, film.title, COUNT(rental.rental_id) AS rental_count,
           @rank := IF(@current_category = category.category_id, @rank + 1, 1) AS ranked,
           @current_category := category.category_id
    FROM rental
    JOIN inventory ON rental.inventory_id = inventory.inventory_id
    JOIN film ON inventory.film_id = film.film_id
    JOIN film_category ON film.film_id = film_category.film_id
    JOIN category ON film_category.category_id = category.category_id
    CROSS JOIN (SELECT @rank := 0, @current_category := NULL) AS vars
    GROUP BY category.category_id, film.title
    ORDER BY category.category_id, rental_count DESC
) AS ranked_films
WHERE ranked <= 3 limit 10;

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[('Action', 'AMERICAN CIRCUS', 22),
 ('Action', 'AMADEUS HOLY', 21),
 ('Action', 'ANTITRUST TOMATOES', 10),
 ('Animation', 'ALTER VICTORY', 22),
 ('Animation', 'ANACONDA CONFESSIONS', 21),
 ('Animation', 'BIKINI BORROWERS', 17),
 ('Children', 'BEAR GRACELAND', 22),
 ('Children', 'BACKLASH UNDEFEATED', 19),
 ('Children', 'BENEATH RUSH', 10),
 ('Classics', 'BOUND CHEAPER', 24)]

In [47]:
%%sql
-- 6. Find the monthly revenue trend for the entire rental store over time.

SELECT YEAR(payment.payment_date) AS year, MONTH(payment.payment_date) AS month, 
       SUM(payment.amount) AS monthly_revenue
FROM payment
GROUP BY YEAR(payment.payment_date), MONTH(payment.payment_date)
ORDER BY year, month;

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(2005, 5, Decimal('4823.44')),
 (2005, 6, Decimal('9629.89')),
 (2005, 7, Decimal('28368.91')),
 (2005, 8, Decimal('24070.14')),
 (2006, 2, Decimal('514.18'))]

In [48]:
%%sql
-- 7. Identify the customers whose total spending on rentals falls within the top 20% of all customers.
    
WITH customer_spending AS (
    SELECT customer_id, SUM(amount) AS total_spent
    FROM payment
    GROUP BY customer_id
),
ranked_customers AS (
    SELECT customer_id, total_spent,
           NTILE(5) OVER (ORDER BY total_spent DESC) AS spending_rank
    FROM customer_spending
)
SELECT customer.customer_id, customer.first_name, customer.last_name, ranked_customers.total_spent
FROM customer
JOIN ranked_customers ON customer.customer_id = ranked_customers.customer_id
WHERE ranked_customers.spending_rank = 1 limit 10;  

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[(526, 'KARL', 'SEAL', Decimal('221.55')),
 (148, 'ELEANOR', 'HUNT', Decimal('216.54')),
 (144, 'CLARA', 'SHAW', Decimal('195.58')),
 (137, 'RHONDA', 'KENNEDY', Decimal('194.61')),
 (178, 'MARION', 'SNYDER', Decimal('194.61')),
 (459, 'TOMMY', 'COLLAZO', Decimal('186.62')),
 (469, 'WESLEY', 'BULL', Decimal('177.60')),
 (468, 'TIM', 'CARY', Decimal('175.61')),
 (236, 'MARCIA', 'DEAN', Decimal('175.58')),
 (181, 'ANA', 'BRADLEY', Decimal('174.66'))]

In [49]:
%%sql
-- 8. Calculate the running total of rentals per category, ordered by rental count.
    
SELECT category.name AS category, film.title, rental.rental_date, 
       COUNT(rental.rental_id) OVER (PARTITION BY category.category_id ORDER BY rental.rental_date) AS running_total
FROM rental
JOIN inventory ON rental.inventory_id = inventory.inventory_id
JOIN film ON inventory.film_id = film.film_id
JOIN film_category ON film.film_id = film_category.film_id
JOIN category ON film_category.category_id = category.category_id
ORDER BY category.name, running_total DESC LIMIT 10;

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[('Action', 'MOCKINGBIRD HOLLYWOOD', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'MIDNIGHT WESTWARD', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'LUST LOCK', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'LUST LOCK', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'LORD ARIZONA', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'GRAIL FRANKENSTEIN', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'DRIFTER COMMANDMENTS', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'DEVIL DESIRE', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'DANCES NONE', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112),
 ('Action', 'DANCES NONE', datetime.datetime(2006, 2, 14, 15, 16, 3), 1112)]

In [50]:
%%sql
-- 9. Find the films that have been rented less than the average rental count for their respective category
WITH film_rentals AS (
    SELECT film_category.category_id, film.film_id, film.title, COUNT(rental.rental_id) AS rental_count
    FROM rental
    JOIN inventory ON rental.inventory_id = inventory.inventory_id
    JOIN film ON inventory.film_id = film.film_id
    JOIN film_category ON film.film_id = film_category.film_id
    GROUP BY film_category.category_id, film.film_id, film.title
),
avg_rental_count AS (
    SELECT category.category_id, AVG(film_rentals.rental_count) AS avg_rentals
    FROM film_rentals
    JOIN category ON film_rentals.category_id = category.category_id
    GROUP BY category.category_id
)
SELECT film_rentals.title, category.name AS category_name, film_rentals.rental_count
FROM film_rentals
JOIN category ON film_rentals.category_id = category.category_id
JOIN avg_rental_count ON film_rentals.category_id = avg_rental_count.category_id
WHERE film_rentals.rental_count < avg_rental_count.avg_rentals limit 10;


 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[('GOSFORD DONNIE', 'Action', 8),
 ('PARK CITIZEN', 'Action', 8),
 ('CASUALTIES ENCINO', 'Action', 9),
 ('MONTEZUMA COMMAND', 'Action', 9),
 ('ANTITRUST TOMATOES', 'Action', 10),
 ('GRAIL FRANKENSTEIN', 'Action', 10),
 ('DARKO DORADO', 'Action', 11),
 ('DRAGON SQUAD', 'Action', 11),
 ('LORD ARIZONA', 'Action', 11),
 ('MAGNOLIA FORRESTER', 'Action', 11)]

In [51]:
%%sql
-- 10. Identify the top 5 months with the highest revenue and display the revenue generated in each month.

SELECT YEAR(payment.payment_date) AS year, MONTH(payment.payment_date) AS month, 
       SUM(payment.amount) AS monthly_revenue
FROM payment
GROUP BY YEAR(payment.payment_date), MONTH(payment.payment_date)
ORDER BY monthly_revenue DESC
LIMIT 5;

 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(2005, 7, Decimal('28368.91')),
 (2005, 8, Decimal('24070.14')),
 (2005, 6, Decimal('9629.89')),
 (2005, 5, Decimal('4823.44')),
 (2006, 2, Decimal('514.18'))]

# Normalisation & CTE

Q 1. To satisfy 1NF, a table must have atomic (indivisible) values, meaning no repeating groups or arrays in a single column.

In the address table, consider if multiple phone numbers could be stored in a single column for each address (e.g., phone = "123-456-7890, 987-654-3210"). This would violate 1NF.

Normalization Solution: Separate phone numbers into their own table, like address_phone, with columns address_id and phone. This way, each phone number is stored in a separate row, achieving atomicity.

Q 2. To be in 2NF, a table must first be in 1NF, and there should be no partial dependencies (i.e., non-key attributes should depend on the whole primary key, not just part of it).

Consider the film_actor table, which has a composite primary key (actor_id, film_id). Suppose this table also includes actor_name. Since actor_name depends only on actor_id (part of the composite key), it violates 2NF.

Normalization Solution: Move actor_name to a separate actor table with actor_id as the primary key. In film_actor, only keep actor_id and film_id.

Q 3. To be in 3NF, a table must be in 2NF and have no transitive dependencies (i.e., non-key attributes should not depend on other non-key attributes).

Consider the film table, which includes columns rental_rate and replacement_cost. If replacement_cost depends on rental_rate or some other calculated rule, this would be a transitive dependency.

Normalization Solution: Move replacement_cost to a separate pricing table that links to the film table by film_id or use a derived column to calculate it when needed.

Question 4 . order_id	customer_id	products	total_cost

1	100	TV, Phone, Laptop	1200
2	101	Phone	500
3	102	Laptop, Phone, Headphones	900
			
1. Unnormalized Form: products column has multiple items.			
			
2. 1NF: Split products into individual rows, moving to a new order_product table.			
			
order_id	product		
1	TV		
1	Phone		
1	Laptop		
2	Phone		
3	Laptop		
3	Phone		
3	Headphones		
			
3. 2NF: Ensure each non-key attribute depends on the whole primary key			


In [52]:
%%sql
WITH actor_films AS (
    SELECT actor.actor_id, actor.first_name, actor.last_name, COUNT(film_actor.film_id) AS film_count
    FROM actor
    JOIN film_actor ON actor.actor_id = film_actor.actor_id
    GROUP BY actor.actor_id
)
SELECT first_name, last_name, film_count
FROM actor_films limit 10;

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[('PENELOPE', 'GUINESS', 19),
 ('NICK', 'WAHLBERG', 25),
 ('ED', 'CHASE', 22),
 ('JENNIFER', 'DAVIS', 22),
 ('JOHNNY', 'LOLLOBRIGIDA', 29),
 ('BETTE', 'NICHOLSON', 20),
 ('GRACE', 'MOSTEL', 30),
 ('MATTHEW', 'JOHANSSON', 20),
 ('JOE', 'SWANK', 25),
 ('CHRISTIAN', 'GABLE', 22)]

In [53]:
%%sql
WITH film_language AS (
    SELECT film.title, language.name AS language, film.rental_rate
    FROM film
    JOIN language ON film.language_id = language.language_id
)
SELECT title, language, rental_rate
FROM film_language limit 10;

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[('ACADEMY DINOSAUR', 'English', Decimal('0.99')),
 ('ACE GOLDFINGER', 'English', Decimal('4.99')),
 ('ADAPTATION HOLES', 'English', Decimal('2.99')),
 ('AFFAIR PREJUDICE', 'English', Decimal('2.99')),
 ('AFRICAN EGG', 'English', Decimal('2.99')),
 ('AGENT TRUMAN', 'English', Decimal('2.99')),
 ('AIRPLANE SIERRA', 'English', Decimal('4.99')),
 ('AIRPORT POLLOCK', 'English', Decimal('4.99')),
 ('ALABAMA DEVIL', 'English', Decimal('2.99')),
 ('ALADDIN CALENDAR', 'English', Decimal('4.99'))]

In [54]:
%%sql
WITH customer_revenue AS (
    SELECT customer_id, SUM(amount) AS total_revenue
    FROM payment
    GROUP BY customer_id
)
SELECT customer_id, total_revenue
FROM customer_revenue limit 10;

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[(1, Decimal('118.68')),
 (2, Decimal('128.73')),
 (3, Decimal('135.74')),
 (4, Decimal('81.78')),
 (5, Decimal('144.62')),
 (6, Decimal('93.72')),
 (7, Decimal('151.67')),
 (8, Decimal('92.76')),
 (9, Decimal('89.77')),
 (10, Decimal('99.75'))]

In [55]:
%%sql
WITH film_ranking AS (
    SELECT title, rental_duration,
           RANK() OVER (ORDER BY rental_duration DESC) AS rank_given
    FROM film
)
SELECT title, rental_duration, rank_given
FROM film_ranking limit 10;

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[('ADAPTATION HOLES', 7, 1),
 ('ANONYMOUS HUMAN', 7, 1),
 ('ARGONAUTS TOWN', 7, 1),
 ('BIKINI BORROWERS', 7, 1),
 ('BLACKOUT PRIVATE', 7, 1),
 ('BLANKET BEVERLY', 7, 1),
 ('BOILED DARES', 7, 1),
 ('BOONDOCK BALLROOM', 7, 1),
 ('BORN SPINAL', 7, 1),
 ('BORROWERS BEDAZZLED', 7, 1)]

In [56]:
%%sql
WITH frequent_customers AS (
    SELECT customer_id, COUNT(rental_id) AS rental_count
    FROM rental
    GROUP BY customer_id
    HAVING rental_count > 2
)
SELECT customer.customer_id, customer.first_name, customer.last_name, frequent_customers.rental_count
FROM customer
JOIN frequent_customers ON customer.customer_id = frequent_customers.customer_id limit 10;

 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[(1, 'MARY', 'SMITH', 32),
 (2, 'PATRICIA', 'JOHNSON', 27),
 (3, 'LINDA', 'WILLIAMS', 26),
 (4, 'BARBARA', 'JONES', 22),
 (5, 'ELIZABETH', 'BROWN', 38),
 (6, 'JENNIFER', 'DAVIS', 28),
 (7, 'MARIA', 'MILLER', 33),
 (8, 'SUSAN', 'WILSON', 24),
 (9, 'MARGARET', 'MOORE', 23),
 (10, 'DOROTHY', 'TAYLOR', 25)]

In [57]:
%%sql
WITH monthly_rentals AS (
    SELECT 
        YEAR(rental_date) AS rental_year,
        MONTH(rental_date) AS rental_month,
        COUNT(rental_id) AS total_rentals
    FROM rental
    GROUP BY YEAR(rental_date), MONTH(rental_date)
)
SELECT rental_year, rental_month, total_rentals
FROM monthly_rentals
ORDER BY rental_year, rental_month;


 * mysql+pymysql://root:***@localhost/sakila
5 rows affected.


[(2005, 5, 1156),
 (2005, 6, 2311),
 (2005, 7, 6709),
 (2005, 8, 5686),
 (2006, 2, 182)]

In [58]:
%%sql
WITH actor_pairs AS (
    SELECT 
        fa1.actor_id AS actor_id1,
        fa2.actor_id AS actor_id2,
        fa1.film_id
    FROM film_actor fa1
    JOIN film_actor fa2 ON fa1.film_id = fa2.film_id AND fa1.actor_id < fa2.actor_id
)
SELECT 
    a1.first_name AS actor1_first_name,
    a1.last_name AS actor1_last_name,
    a2.first_name AS actor2_first_name,
    a2.last_name AS actor2_last_name,
    actor_pairs.film_id
FROM actor_pairs
JOIN actor a1 ON actor_pairs.actor_id1 = a1.actor_id
JOIN actor a2 ON actor_pairs.actor_id2 = a2.actor_id limit 10;


 * mysql+pymysql://root:***@localhost/sakila
10 rows affected.


[('PENELOPE', 'GUINESS', 'CHRISTIAN', 'GABLE', 1),
 ('PENELOPE', 'GUINESS', 'LUCILLE', 'TRACY', 1),
 ('PENELOPE', 'GUINESS', 'SANDRA', 'PECK', 1),
 ('PENELOPE', 'GUINESS', 'JOHNNY', 'CAGE', 1),
 ('PENELOPE', 'GUINESS', 'MENA', 'TEMPLE', 1),
 ('PENELOPE', 'GUINESS', 'WARREN', 'NOLTE', 1),
 ('PENELOPE', 'GUINESS', 'OPRAH', 'KILMER', 1),
 ('PENELOPE', 'GUINESS', 'ROCK', 'DUKAKIS', 1),
 ('PENELOPE', 'GUINESS', 'MARY', 'KEITEL', 1),
 ('PENELOPE', 'GUINESS', 'JENNIFER', 'DAVIS', 23)]

In [59]:
%%sql
--  Implement a recursive CTE to find all employees in the staff table who report to a specific manager, considering the reports_to column

WITH RECURSIVE employee_hierarchy AS (
    -- Anchor member: select the specific manager
    SELECT staff_id, first_name, last_name
    FROM staff
    WHERE staff_id = 1  

    UNION ALL

    SELECT s.staff_id, s.first_name, s.last_name
    FROM staff s
    JOIN employee_hierarchy eh ON s.address_id = eh.staff_id
)

SELECT staff_id, first_name, last_name
FROM employee_hierarchy;


 * mysql+pymysql://root:***@localhost/sakila
1 rows affected.


[(1, 'Mike', 'Hillyer')]